tket/src/WeightSubgrMono/EndToEndWrappers/MainSolver.cpp

```diff
@@ -0,0 +1,421 @@
+          m_target_neighbours_data,
+          m_pre_search_components_ptr->target_near_ndata,
+          parameters.max_distance_for_distance_reduction_during_search,
+          m_solution_data.extra_statistics);
+    }
+  }
+
+  m_solution_data.initialisation_time_ms =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          Clock::now() - init_start)
+          .count();
+
+  if (!initialisation_succeeded) {
+    m_solution_data.finished = true;
+    return;
+  }
+
+  // The problem is not obviously impossible, so fill in more information.
+  {
+    auto p_weights = m_pattern_neighbours_data.get_weights_expensive();
+    std::sort(p_weights.begin(), p_weights.end());
+
+    auto t_weights = m_target_neighbours_data.get_weights_expensive();
+    std::sort(t_weights.begin(), t_weights.end());
+
+    TKET_ASSERT(
+        p_weights.size() == m_pattern_neighbours_data.get_number_of_edges());
+    TKET_ASSERT(
+        t_weights.size() == m_target_neighbours_data.get_number_of_edges());
+    TKET_ASSERT(p_weights.size() <= t_weights.size());
+    m_solution_data.total_p_edge_weights =
+        std::accumulate(p_weights.cbegin(), p_weights.cend(), WeightWSM(0));
+
+    // Now get trivial lower, upper bounds.
+    // When considering   sum a[i].b[p(i)],  where p can be any permutation
+    // and a,b are sequences of real numbers,
+    // the min value arises when p is such that a,b are in opposite order.
+    // The max value arises when a,b are in the same order.
+    m_solution_data.trivial_weight_lower_bound = 0;
+
+    // Lower bound: p-weights increasing, t-weights decreasing;
+    // BUT we only need use the smallest t-weights, for the minimum.
+    for (unsigned ii = 0; ii < p_weights.size(); ++ii) {
+      const auto product = get_product_or_throw(
+          p_weights[ii], t_weights[(p_weights.size() - 1) - ii]);
+      m_solution_data.trivial_weight_lower_bound =
+          get_sum_or_throw(m_solution_data.trivial_weight_lower_bound, product);
+    }
+
+    // Now, for the MAXIMUM, we use the LARGEST t-weights,
+    // as well as summing both vectors in increasing order.
+    const unsigned offset = t_weights.size() - p_weights.size();
+    m_solution_data.trivial_weight_initial_upper_bound = 0;
+
+    for (unsigned ii = 0; ii < p_weights.size(); ++ii) {
+      const auto product =
+          get_product_or_throw(p_weights[ii], t_weights[ii + offset]);
+      m_solution_data.trivial_weight_initial_upper_bound = get_sum_or_throw(
+          m_solution_data.trivial_weight_initial_upper_bound, product);
+    }
+  }
+
+  if (m_solution_data.trivial_weight_lower_bound !=
+      m_solution_data.trivial_weight_initial_upper_bound) {
+    // It's not an unweighted problem, so it's worth checking for weights.
+    m_search_branch_ptr->activate_weight_checker(
+        m_solution_data.total_p_edge_weights);
+  }
+
+  if (m_solution_data.initialisation_time_ms >= parameters.timeout_ms) {
```

Comment:
This is a bit silly, but:

There is a middle bit of calculation between "initialisation" and beginning the proper search which it's nice to fill in, but actually should take negligible time compared with initialisation. But yes, strictly speaking I should recalculate it after the middle bit. I was worried about Clock::now() calls themselves taking significant time, but I suppose they can't, otherwise they'd be quite silly.

(Or, in other words, we really have "initialise part 1" and "part 2", and it drops out early if part 1 fails, and returns the part 1 time; but one should really add on the part 2 time; even though in practice it SHOULD always be negligible).